storage/indexer/src/db_tailer.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    schema::{
+        event_by_key::EventByKeySchema, event_by_version::EventByVersionSchema,
+        indexer_metadata::TailerMetadataSchema, transaction_by_account::TransactionByAccountSchema,
+    },
+    utils::{
+        error_if_too_many_requested, get_first_seq_num_and_limit, AccountTransactionVersionIter,
+        MAX_REQUEST_LIMIT,
+    },
+};
+use aptos_config::config::index_db_tailer_config::IndexDBTailerConfig;
+use aptos_schemadb::{ReadOptions, SchemaBatch, DB};
+use aptos_storage_interface::{
+    db_ensure as ensure, db_other_bail as bail, AptosDbError, DbReader, Result,
+};
+use aptos_types::{
+    account_address::AccountAddress,
+    contract_event::{ContractEvent, EventWithVersion},
+    event::EventKey,
+    indexer::db_tailer_reader::{IndexerTransactionEventReader, Order},
+    transaction::{AccountTransactionsWithProof, Version},
+};
+use std::sync::Arc;
+
+pub struct DBTailer {
+    pub db: Arc<DB>,
+    pub main_db_reader: Arc<dyn DbReader>,
+    batch_size: usize,
+}
+
+impl DBTailer {
+    pub fn new(db: Arc<DB>, db_reader: Arc<dyn DbReader>, config: &IndexDBTailerConfig) -> Self {
+        Self {
+            db,
+            main_db_reader: db_reader,
+            batch_size: config.batch_size,
+        }
+    }
+
+    pub fn get_persisted_version(&self) -> Version {
+        // read the latest key from the db
+        let mut rev_iter_res = self
+            .db
+            .rev_iter::<TailerMetadataSchema>(Default::default())
+            .expect("Cannot create db tailer metadata iterator");
+        rev_iter_res
+            .next()
+            .map(|res| res.map_or(0, |(version, _)| version))
+            .unwrap_or_default()
+    }
+
+    pub fn process_a_batch(&self, start_version: Option<Version>) -> Result<Version> {
+        let mut version = start_version.unwrap_or(0);
+        let db_iter: Box<
+            dyn Iterator<
+                Item = std::prelude::v1::Result<
+                    (aptos_types::transaction::Transaction, Vec<ContractEvent>),
+                    AptosDbError,
+                >,
+            >,
+        > = self
+            .main_db_reader
+            .get_db_backup_iter(version, self.batch_size)
+            .expect("Cannot create db tailer iterator");
+        let batch = SchemaBatch::new();
+        let metadata_batch = SchemaBatch::new();
+        db_iter.for_each(|res| {
```

Comment:
nit: consider using rayon here.

Reply:
don't know how to do this? this db_iter is sequential itself.